tiledb/sm/filesystem/s3.cc

```diff
@@ -255,58 +267,83 @@ Status S3::flush_object(const URI& uri) {
-    return LOG_STATUS(Status::S3Error(
-        "Unable to find completed parts list for S3 object " +
-        uri.to_string()));
+  if (state.st.ok()) {
+    Aws::S3::Model::CompleteMultipartUploadRequest complete_request =
+        build_multipart_upload_request(state);
+    auto outcome = client_->CompleteMultipartUpload(complete_request);
+
+    return finish_flush_object(
+        std::move(outcome), uri, std::move(multipart_lck), state, buff);
+  } else {
+    Aws::S3::Model::AbortMultipartUploadRequest abort_request =
+        build_multipart_abort_request(state);
+    auto outcome = client_->AbortMultipartUpload(abort_request);
+
+    return finish_flush_object(
+        std::move(outcome), uri, std::move(multipart_lck), state, buff);
   }
-  for (auto& tup : completed_parts_it->second) {
-    Aws::S3::Model::CompletedPart& part = std::get<1>(tup);
-    completed_multipart_upload.AddParts(part);
+}
+
+Aws::S3::Model::CompleteMultipartUploadRequest
+S3::build_multipart_upload_request(const MultiPartUploadState& state) {
+  // Add all the completed parts (sorted by part number) to the upload object.
+  Aws::S3::Model::CompletedMultipartUpload completed_upload;
+  for (auto& tup : state.completed_parts) {
+    const Aws::S3::Model::CompletedPart& part = std::get<1>(tup);
+    completed_upload.AddParts(part);
   }
 
-  auto multipart_upload_request_it = multipart_upload_request_.find(path_c_str);
-  auto complete_multipart_upload_request = multipart_upload_request_it->second;
-  complete_multipart_upload_request.WithMultipartUpload(
-      completed_multipart_upload);
-  auto complete_multipart_upload_outcome =
-      client_->CompleteMultipartUpload(complete_multipart_upload_request);
+  Aws::S3::Model::CompleteMultipartUploadRequest complete_request;
+  complete_request.SetBucket(state.bucket);
+  complete_request.SetKey(state.key);
+  complete_request.SetUploadId(state.upload_id);
+  return complete_request.WithMultipartUpload(std::move(completed_upload));
+}
+
+Aws::S3::Model::AbortMultipartUploadRequest S3::build_multipart_abort_request(
+    const MultiPartUploadState& state) {
+  Aws::S3::Model::AbortMultipartUploadRequest abort_request;
+  abort_request.SetBucket(state.bucket);
+  abort_request.SetKey(state.key);
+  abort_request.SetUploadId(state.upload_id);
+  return abort_request;
+}
 
-  // Release lock while we wait for the file to flush.
+template <typename R, typename E>
+Status S3::finish_flush_object(
+    const Aws::Utils::Outcome<R, E>& outcome,
+    const URI& uri,
+    std::unique_lock<std::mutex>&& multipart_lck,
+    const MultiPartUploadState& state,
+    Buffer* const buff) {
+  auto bucket = state.bucket;
+  auto key = state.key;
   multipart_lck.unlock();
 
-  wait_for_object_to_propagate(
-      complete_multipart_upload_request.GetBucket(),
-      complete_multipart_upload_request.GetKey());
+  wait_for_object_to_propagate(move(bucket), move(key));
```

Comment:
Is this correct in the error case? If some part failed to upload, and we end up here on the abort path, I think this wait will block needlessly until the max number of retries is reached (as the object may never be created in that case).

Reply:
That makes more sense, I'll delete this line.

Reply:
I don't think you want to remove it altogether, it's still necessary on the non-error path, in order to retain the semantics we had before.

Reply:
Agreed, sorry I was not clear: I'll remove this line from the abort path.